src/java/main/org/apache/zookeeper/server/NIOServerCnxnFactory.java

```diff
@@ -842,7 +868,26 @@ private void addCnxn(NIOServerCnxn cnxn) {
 
     protected NIOServerCnxn createConnection(SocketChannel sock,
             SelectionKey sk, SelectorThread selectorThread) throws IOException {
-        return new NIOServerCnxn(zkServer, sock, sk, this, selectorThread);
+
+        NIOServerCnxn cnxn = null;
+
+        if (serverCnxnClassCtr != null) {
```

Comment:
Actually I tend to come back to double instantiation - as it will be used only in UnitTests (serverCnxnClassCtr != null), while manipulation with constructor (invocation of newInstance) instead of directly creating instance of NIOServerCnxn will be really a hot path